Sources/ArcGISToolkit/FloorFilter/FloorFilterViewModel.swift

```diff
@@ -0,0 +1,118 @@
+    public var visibleLevelsInExpandedList: [AGSFloorLevel] {
+        guard let floorManager = floorManager else { return [] }
+        return facilities.isEmpty ? floorManager.levels : floorManager.levels.filter { $0.facility == selectedFacility }.reversed()
+    }
+    
+    // All the levels in the map
+    public var allLevels: [AGSFloorLevel] {
+        return floorManager?.levels ?? []
+    }
+    
+    // The site, facility, and level that are selected by the user
+    public var selectedSite: AGSFloorSite?
+    public var selectedFacility: AGSFloorFacility?
+    public var selectedLevel: AGSFloorLevel?
+    
+    // The default vertical order is 0 according to Runtime 100.12 update for AGSFloorManager
+    public let defaultVerticalOrder = 0
+
+    public func reset() {
+        floorManager = nil
+        selectedSite = nil
+        selectedFacility = nil
+        selectedLevel = nil
+    }
+    
+    public func getSelectedSite() -> AGSFloorSite? {
+        return sites.first { $0 == selectedSite }
+    }
+
+    public func getSelectedFacility() -> AGSFloorFacility? {
+        return facilities.first { $0 == selectedFacility }
+    }
+    
+    public func getSelectedVisibleLevel() -> AGSFloorLevel? {
+        return visibleLevelsInExpandedList.first { $0 == selectedLevel }
+    }
+    
+    // Sets the visibility of all the levels on the map based on the vertical order of the current selected level
+    public func filterMapToSelectedLevel() {
+        guard let selectedLevel = getSelectedVisibleLevel() ?? selectedLevel else { return }
+        allLevels.forEach {
+            $0.isVisible = $0.verticalOrder == selectedLevel.verticalOrder
+        }
+    }
+    
+    // Zooms to the facility if there is a selected facility, otherwise zooms to the site if there is no selected facility
+    public func zoomToSelection() {
+        if let _ = selectedFacility {
+            zoomToFacility()
+        } else {
+            if let _ = selectedSite {
+                zoomToSite()
+            }
+        }
+    }
+    
+    private func zoomToSite() {
+        zoomToExtent(mapView: mapView, envelope: getSelectedSite()?.geometry?.extent)
+    }
+    
+    private func zoomToFacility() {
+        zoomToExtent(mapView: mapView, envelope: getSelectedFacility()?.geometry?.extent)
+    }
+    
+    private func zoomToExtent(mapView: AGSMapView?, envelope: AGSEnvelope?, padding: Double = 1.5) {
+        if let mapView = mapView, let envelope = envelope {
+            let envelopeWithBuffer = AGSEnvelope(center: envelope.center, width: envelope.width * padding, height: envelope.height * padding)
+            if (!envelopeWithBuffer.isEmpty) {
+                let viewPoint = AGSViewpoint(targetExtent: envelopeWithBuffer)
+                DispatchQueue.main.async {
+                    mapView.setViewpoint(viewPoint, duration: 0.5, completion: nil)
+                }
+            }
+            
+        }
+    }
```

Comment:
Just a few changes with this method...
I generally like optional arguments, but if it's never used, it should be removed and placed in the func.
A `guard` statement is best here and helps readability.
Reformatted the code creating the envelope for readability
Don't need parenthesis around the last `if`. (I still do that sometimes ;-) )
Because `mapView.setViewpoint(...)` is being called in response to a user interaction (i.e., from the main thread), it doesn't need the `DispatchQueue.main.async` around it.

```suggestion
    private func zoomToExtent(mapView: AGSMapView?, envelope: AGSEnvelope?) {
        guard let mapView = mapView,
              let envelope = envelope
        else { return }
        
        let padding = 1.5
        let envelopeWithBuffer = AGSEnvelope(
            center: envelope.center,
            width: envelope.width * padding,
            height: envelope.height * padding
        )
        
        if !envelopeWithBuffer.isEmpty {
            let viewPoint = AGSViewpoint(targetExtent: envelopeWithBuffer)
            mapView.setViewpoint(viewPoint, duration: 0.5, completion: nil)
        }
    }
```